File: tools/tune.cpp
#include <algorithm>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <regex>
#include <string>
#include <unordered_map>
#include <vector>

#include "lilia/constants.hpp"
#include "lilia/model/chess_game.hpp"

using namespace lilia;

struct Parameters {
  std::unordered_map<std::string, double> values;

  static Parameters load(const std::filesystem::path &path =
                             "include/lilia/engine/eval_tune_shared.hpp") {
    Parameters p;
    std::ifstream in(path);
    if (!in) {
      std::cerr << "Failed to open " << path << "\n";
      return p;
    }
    std::string content((std::istreambuf_iterator<char>(in)), {});
    std::regex pattern(R"(inline\s+int\s+(\w+)\s*=\s*([-0-9]+))");
    std::smatch m;
    auto begin = content.cbegin();
    while (std::regex_search(begin, content.cend(), m, pattern)) {
      p.values[m[1]] = std::stod(m[2]);
      begin = m.suffix().first;
    }
    return p;
  }

  void save_as_constexpr(const std::filesystem::path &out_path) const {
    std::ofstream out(out_path);
    for (const auto &kv : values) {
      out << "constexpr int " << kv.first << " = "
          << static_cast<int>(std::round(kv.second)) << ";\n";
    }
  }
};

double cp_to_wdl(double cp) { return 1.0 / (1.0 + std::exp(-cp / 400.0)); }

double evaluate_board(const std::string &fen, const Parameters &params) {
  const auto &p = params.values;
  double piece_values[6];
  piece_values[0] = p.at("PAWN_VALUE");
  piece_values[1] = p.at("KNIGHT_VALUE");
  piece_values[2] = p.at("BISHOP_VALUE");
  piece_values[3] = p.at("ROOK_VALUE");
  piece_values[4] = p.at("QUEEN_VALUE");
  piece_values[5] = 0.0; // King

  int counts[2][6] = {};
  std::string placement = fen.substr(0, fen.find(' '));
  for (char ch : placement) {
    if (ch == '/')
      continue;
    if (std::isdigit(static_cast<unsigned char>(ch))) {
      continue; // skip empty squares
    }
    bool white = std::isupper(static_cast<unsigned char>(ch));
    char piece = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
    int idx = 5;
    switch (piece) {
    case 'p':
      idx = 0;
      break;
    case 'n':
      idx = 1;
      break;
    case 'b':
      idx = 2;
      break;
    case 'r':
      idx = 3;
      break;
    case 'q':
      idx = 4;
      break;
    case 'k':
    default:
      idx = 5;
      break;
    }
    counts[white ? 0 : 1][idx]++;
  }

  double score = 0.0;
  for (int i = 0; i < 5; ++i)
    score += piece_values[i] * (counts[0][i] - counts[1][i]);

  bool white_turn = fen.find(" w ") != std::string::npos;
  score += white_turn ? p.at("TEMPO_MG") : -p.at("TEMPO_MG");
  if (counts[0][2] >= 2)
    score += p.at("BISHOP_PAIR");
  if (counts[1][2] >= 2)
    score -= p.at("BISHOP_PAIR");
  return score;
}

std::vector<std::string> generate_fens(int samples) {
  model::ChessGame game;
  std::vector<std::string> fens;
  std::mt19937 rng(std::random_device{}());
  while (static_cast<int>(fens.size()) < samples) {
    const auto &moves = game.generateLegalMoves();
    if (moves.empty()) {
      game.setPosition(core::START_FEN);
      continue;
    }
    std::uniform_int_distribution<size_t> dist(0, moves.size() - 1);
    const auto &m = moves[dist(rng)];
    game.doMove(m.from(), m.to(), m.promotion());
    game.checkGameResult();
    fens.push_back(game.getFen());
    if (game.getResult() != core::GameResult::ONGOING) {
      game.setPosition(core::START_FEN);
    }
  }
  return fens;
}

std::vector<double> label_fens(const std::vector<std::string> &fens,
                               const Parameters &params) {
  std::vector<double> labels;
  labels.reserve(fens.size());
  for (const auto &fen : fens) {
    double cp = evaluate_board(fen, params);
    labels.push_back(cp_to_wdl(cp));
  }
  return labels;
}

double texel_loss(const std::vector<double> &pred,
                  const std::vector<double> &target) {
  double loss = 0.0;
  const double eps = 1e-12;
  for (size_t i = 0; i < pred.size(); ++i) {
    loss += target[i] * std::log(pred[i] + eps) +
            (1.0 - target[i]) * std::log(1.0 - pred[i] + eps);
  }
  return -loss / pred.size();
}

Parameters spsa_optimize(const Parameters &start_params,
                         const std::vector<std::string> &fens,
                         const std::vector<double> &labels, int iterations = 100,
                         double a = 0.1, double c = 0.1) {
  std::vector<std::string> keys;
  std::vector<double> theta;
  keys.reserve(start_params.values.size());
  theta.reserve(start_params.values.size());
  for (const auto &kv : start_params.values) {
    keys.push_back(kv.first);
    theta.push_back(kv.second);
  }
  std::mt19937 rng(std::random_device{}());

  auto evaluate = [&](const std::vector<double> &theta_vec) {
    Parameters p;
    for (size_t i = 0; i < keys.size(); ++i)
      p.values[keys[i]] = theta_vec[i];
    std::vector<double> preds;
    preds.reserve(fens.size());
    for (const auto &fen : fens) {
      double cp = evaluate_board(fen, p);
      preds.push_back(cp_to_wdl(cp));
    }
    return texel_loss(preds, labels);
  };

  for (int k = 1; k <= iterations; ++k) {
    double alpha = a / std::pow(k, 0.602);
    double gamma = c / std::pow(k, 0.101);
    std::vector<double> delta(theta.size());
    for (auto &d : delta)
      d = std::uniform_int_distribution<int>(0, 1)(rng) ? 1.0 : -1.0;
    std::vector<double> theta_plus(theta.size()), theta_minus(theta.size());
    for (size_t i = 0; i < theta.size(); ++i) {
      theta_plus[i] = theta[i] + gamma * delta[i];
      theta_minus[i] = theta[i] - gamma * delta[i];
    }
    double loss_plus = evaluate(theta_plus);
    double loss_minus = evaluate(theta_minus);
    for (size_t i = 0; i < theta.size(); ++i) {
      double grad = (loss_plus - loss_minus) / (2.0 * gamma * delta[i]);
      theta[i] -= alpha * grad;
    }
  }
  Parameters tuned;
  for (size_t i = 0; i < keys.size(); ++i)
    tuned.values[keys[i]] = theta[i];
  return tuned;
}

int main(int argc, char **argv) {
  int samples = 100;
  int iterations = 50;
  std::string export_path;
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--samples" && i + 1 < argc) {
      samples = std::stoi(argv[++i]);
    } else if (arg == "--iterations" && i + 1 < argc) {
      iterations = std::stoi(argv[++i]);
    } else if (arg == "--export" && i + 1 < argc) {
      export_path = argv[++i];
    }
  }

  Parameters params = Parameters::load();
  auto fens = generate_fens(samples);
  auto labels = label_fens(fens, params);
  auto tuned = spsa_optimize(params, fens, labels, iterations);

  if (!export_path.empty()) {
    tuned.save_as_constexpr(export_path);
    std::cout << "Exported tuned parameters to " << export_path << "\n";
  } else {
    for (const auto &kv : tuned.values) {
      std::cout << kv.first << " = " << std::fixed << std::setprecision(2)
                << kv.second << "\n";
    }
  }
  return 0;
}

File: tools/tune.py
#!/usr/bin/env python3
"""End-to-end evaluation parameter tuner.

This script generates training data via Stockfish self play, labels positions
using Stockfish evaluations converted to win/draw/loss probabilities and tunes
engine evaluation parameters using Texel loss with a simple SPSA optimiser.

The tunable parameters live in ``include/lilia/engine/eval_tune_shared.hpp``.
After optimisation the resulting parameters can be written back as ``constexpr``
values to be used for normal engine builds.

Note: The default configuration is intentionally conservative so that the script
can be executed in a restricted environment.  For large scale tuning (1M+
positions) increase the ``--samples`` argument accordingly.
"""

from __future__ import annotations

import argparse
import math
import os
import random
import re
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, List, Tuple

import numpy as np
import chess
import chess.engine

# ---------------------------------------------------------------------------
# Helper functions
# ---------------------------------------------------------------------------

HEADER_PATH = Path(__file__).resolve().parent.parent / "include" / "lilia" / "engine" / "eval_tune_shared.hpp"


@dataclass
class Parameters:
    values: Dict[str, float]

    @classmethod
    def load(cls) -> "Parameters":
        text = HEADER_PATH.read_text(encoding="utf-8")
        pattern = re.compile(r"inline\s+int\s+(\w+)\s*=\s*([-0-9]+)")
        values = {m.group(1): float(m.group(2)) for m in pattern.finditer(text)}
        return cls(values)

    def save_as_constexpr(self, out_path: Path) -> None:
        content_lines = [f"constexpr int {k} = {int(round(v))};" for k, v in self.values.items()]
        out_path.write_text("\n".join(content_lines) + "\n", encoding="utf-8")


# Texel formula converting centipawns to win probability.
def cp_to_wdl(cp: float) -> float:
    return 1.0 / (1.0 + math.exp(-cp / 400.0))


# Simple material based evaluation using a subset of parameters.
def evaluate_board(board: chess.Board, params: Parameters) -> float:
    p = params.values
    piece_values = {
        chess.PAWN: p.get("PAWN_VALUE", 100),
        chess.KNIGHT: p.get("KNIGHT_VALUE", 320),
        chess.BISHOP: p.get("BISHOP_VALUE", 330),
        chess.ROOK: p.get("ROOK_VALUE", 500),
        chess.QUEEN: p.get("QUEEN_VALUE", 900),
        chess.KING: 0,
    }
    score = 0.0
    for piece_type in piece_values:
        score += piece_values[piece_type] * (
            len(board.pieces(piece_type, chess.WHITE))
            - len(board.pieces(piece_type, chess.BLACK))
        )
    # Tempo bonus
    score += p.get("TEMPO_MG", 0) if board.turn == chess.WHITE else -p.get("TEMPO_MG", 0)
    # Bishop pair
    if len(board.pieces(chess.BISHOP, chess.WHITE)) >= 2:
        score += p.get("BISHOP_PAIR", 0)
    if len(board.pieces(chess.BISHOP, chess.BLACK)) >= 2:
        score -= p.get("BISHOP_PAIR", 0)
    return score


# Generate FEN positions via Stockfish self play.
def generate_fens(engine_path: str, samples: int) -> List[str]:
    engine = chess.engine.SimpleEngine.popen_uci(engine_path)
    fens = []
    board = chess.Board()
    while len(fens) < samples:
        result = engine.play(board, chess.engine.Limit(depth=1))
        board.push(result.move)
        fens.append(board.fen())
        if board.is_game_over():
            board.reset()
    engine.quit()
    return fens


# Label positions with Stockfish evaluation converted to WDL probability.
def label_fens(engine_path: str, fens: Iterable[str]) -> List[float]:
    engine = chess.engine.SimpleEngine.popen_uci(engine_path)
    labels = []
    for fen in fens:
        board = chess.Board(fen)
        info = engine.analyse(board, chess.engine.Limit(depth=10))
        cp = info["score"].white().score(mate_score=10000)
        labels.append(cp_to_wdl(float(cp)))
    engine.quit()
    return labels


# Texel loss (logistic cross entropy).
def texel_loss(pred: np.ndarray, target: np.ndarray) -> float:
    eps = 1e-12
    return float(-np.mean(target * np.log(pred + eps) + (1 - target) * np.log(1 - pred + eps)))


# SPSA optimiser.
def spsa_optimize(params: Parameters, fens: List[str], labels: List[float],
                  iterations: int = 100, a: float = 0.1, c: float = 0.1) -> Parameters:
    keys = list(params.values.keys())
    theta = np.array([params.values[k] for k in keys], dtype=float)
    labels_np = np.array(labels, dtype=float)

    def evaluate(theta_vec: np.ndarray) -> float:
        p_dict = {k: v for k, v in zip(keys, theta_vec)}
        local_params = Parameters(p_dict)
        preds = [cp_to_wdl(evaluate_board(chess.Board(f), local_params)) for f in fens]
        return texel_loss(np.array(preds), labels_np)

    for k in range(1, iterations + 1):
        alpha = a / (k ** 0.602)
        gamma = c / (k ** 0.101)
        delta = np.array([random.choice([-1, 1]) for _ in theta])
        loss_plus = evaluate(theta + gamma * delta)
        loss_minus = evaluate(theta - gamma * delta)
        grad = (loss_plus - loss_minus) / (2.0 * gamma * delta)
        theta -= alpha * grad
    tuned = Parameters({k: v for k, v in zip(keys, theta)})
    return tuned


# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(description="End-to-end SPSA tuner")
    parser.add_argument("--stockfish", default="stockfish", help="Path to Stockfish binary")
    parser.add_argument("--samples", type=int, default=100, help="Number of FEN samples to generate")
    parser.add_argument("--iterations", type=int, default=50, help="Number of SPSA iterations")
    parser.add_argument("--export", type=Path, help="Export tuned parameters to header")
    args = parser.parse_args()

    params = Parameters.load()
    fens = generate_fens(args.stockfish, args.samples)
    labels = label_fens(args.stockfish, fens)
    tuned = spsa_optimize(params, fens, labels, iterations=args.iterations)

    if args.export:
        tuned.save_as_constexpr(args.export)
        print(f"Exported tuned parameters to {args.export}")
    else:
        for k, v in tuned.values.items():
            print(f"{k} = {v:.2f}")


if __name__ == "__main__":
    main()
